cpp/src/arrow/compute/kernels/mask.h

```diff
@@ -0,0 +1,82 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+
+#include "arrow/compute/kernel.h"
+#include "arrow/status.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+class Array;
+
+namespace compute {
+
+class FunctionContext;
+
+struct ARROW_EXPORT MaskOptions {};
+
+/// \brief Mask an array with a boolean selection filter
+///
+/// The output array will be populated with values from the input at positions
+/// where the selection filter is not 0. Nulls in the filter will result in nulls
+/// in the output.
+///
+/// For example given values = ["a", "b", "c", null, "e", "f"] and
+/// filter = [0, 1, 1, 0, null, 1], the output will be
+/// = ["b", "c", null, "f"]
+///
+/// \param[in] context the FunctionContext
+/// \param[in] values array from which to take
+/// \param[in] mask indicates which values should be masked out
+/// \param[in] options options
+/// \param[out] out resulting array
+ARROW_EXPORT
+Status Mask(FunctionContext* context, const Array& values, const Array& mask,
```

Comment:
Shouldn't this be called "Select" or "Filter"?

Reply:
Unfortunately `Filter` is [already taken](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/filter.h#L34). Select would be fine, anyone else want to weigh in?

Reply:
I wanted to give the same comment actually, but not sure it was nitpicking.

The doc comment now basically says "Mask an array with a boolean filter", while I would typically speak about "Filter an array with a boolean mask".

Reply:
If it is not too late to rename `Filter`, I would rather call that Mask and the one in this PR Filter.

(although Mask is maybe also not ideal for the current Filter, as it does not "mask" (as a verb) but it "creates a mask")

Reply:
@fsaintjacques Maybe we could rename `FilterFunction -> CompareFunction`, then `Mask -> Filter`

Reply:
I have no objection with the rename, just make sure to update the glib bindings.